ambry-router/src/main/java/com/github/ambry/router/SimpleOperationTracker.java

```diff
@@ -419,6 +424,48 @@ public boolean maybeFailedDueToOfflineReplicas() {
     if (!routerConfig.routerUnavailableDueToOfflineReplicas) {
       return false;
     }
+
+    // For GetBlob
+    if (routerOperation == RouterOperation.GetBlobOperation || routerOperation == RouterOperation.GetBlobInfoOperation) {
+      // Explanation of the following three conditions in order,
+      // 1. hasFailedOnCrossColoNotFound:
+      //    We tried all the "eligible" replicas at this moment, couldn't finish the operation due to NOT_FOUND
+      // 2. (allReplicaCount - totalNotFoundCount >= replicaSuccessTarget)
+      //    Minus all the NOT_FOUND replicas, among all the replicas belongs to the partition(no matter online or offline),
+      //    we should still have enough replicas to satisfy the replicaSuccessTarget.
+      //    That means, if we retry, potentially we have the chance to be successful and get the blob back from other replica.
+      // 3. (originatingDcEligibleReplicaCount < originatingDcTotalReplicaCount - routerConfig.routerPutSuccessTarget + 1)
+      //    The retry can be successful only if there is some offline replica coming back when retry happens.
+      //    That's why for TtlUpdate and Delete, we have the third condition as (totalOfflineReplicaCount > 0)
+      //    a). For GetBlob, the difference is that we already include OFFLINE replicas to the eligible replica pool
+      //        while TTLUpdate/Delete don't. Refer to routerOperationTrackerIncludeDownReplicas related code.
+      //        So it has tried the OFFLINE replicas already.
+      //    b). But with the recent Helix implementation, when the server is crashed,
+      //        the helix external view sometimes doesn't include the replica at all, not even in the OFFLINE category.
+      //        So we check how many replicas from the originating DC we read already.
+      //        Since we put routerPutSuccessTarget copies in the originating DC,
+      //        we should at least read "originatingDcTotalReplicaCount - routerConfig.routerPutSuccessTarget + 1" replicas
+      //        to make sure we read one of the copies.
+      //        If the eligible originating replica count is less that that, we should retry.
+      if (hasFailedOnCrossColoNotFound()
+          && (allReplicaCount - totalNotFoundCount >= replicaSuccessTarget)
+          && (originatingDcEligibleReplicaCount < originatingDcTotalReplicaCount - routerConfig.routerPutSuccessTarget + 1)) {
```

Comment:
I think you probably have a typo. Do you mean "isn't originatingDcEligibleReplicaCount the same as 
 originatingDcTotalReplicaCount"?
If so, it's not.
originatingDcEligibleReplicaCount is the originating DC in the eligible pool. It can be 1, 2 or 3.

originatingDcTotalReplicaCount is the total originating DC replica count which is 3.